array_api_tests/dtype_helpers.py

```diff
@@ -90,31 +133,37 @@ class MinMax(NamedTuple):
     max: Union[int, float]
 
 
-dtype_ranges = {
-    xp.int8: MinMax(-128, +127),
-    xp.int16: MinMax(-32_768, +32_767),
-    xp.int32: MinMax(-2_147_483_648, +2_147_483_647),
-    xp.int64: MinMax(-9_223_372_036_854_775_808, +9_223_372_036_854_775_807),
-    xp.uint8: MinMax(0, +255),
-    xp.uint16: MinMax(0, +65_535),
-    xp.uint32: MinMax(0, +4_294_967_295),
-    xp.uint64: MinMax(0, +18_446_744_073_709_551_615),
-    xp.float32: MinMax(-3.4028234663852886e+38, 3.4028234663852886e+38),
-    xp.float64: MinMax(-1.7976931348623157e+308, 1.7976931348623157e+308),
-}
-
-dtype_nbits = {
-    **{d: 8 for d in [xp.int8, xp.uint8]},
-    **{d: 16 for d in [xp.int16, xp.uint16]},
-    **{d: 32 for d in [xp.int32, xp.uint32, xp.float32]},
-    **{d: 64 for d in [xp.int64, xp.uint64, xp.float64]},
-}
-
-
-dtype_signed = {
-    **{d: True for d in int_dtypes},
-    **{d: False for d in uint_dtypes},
-}
+dtype_ranges = EqualityMapping(
+    {
+        xp.int8: MinMax(-128, +127),
```

Comment:
These still create the dict. If a dtype object really didn't support hashing it would fail here.

Reply:
Ahh 🤦 I 'spose this solves our immediate problem of supporting NumPy proper. Replacing dicts here are pretty annoying, so we can leave it later as it's not high priority. 

I'll write a top-level assertion to see if all the dtype objects support hashing for now.

Reply:
Should just be a question of using a list of tuples instead of a dictionary, no?

Reply:
> Should just be a question of using a list of tuples instead of a dictionary, no?

Yep, pushed a change which now uses key-value tuples for the internal mapping, and subsequently updated the respective dtype helpers.